lib/semantics/check-select-stmt.cc

```diff
@@ -0,0 +1,239 @@
+//===-- lib/semantics/check-select-stmt.cc --------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "check-select-stmt.h"
+#include "tools.h"
+#include "../parser/message.h"
+
+namespace Fortran::semantics {
+
+void SelectStmtChecker::Leave(
+    const parser::CaseConstruct &selectCaseConstruct) {
+  const auto &selectCaseStmt{
+      std::get<parser::Statement<parser::SelectCaseStmt>>(
+          selectCaseConstruct.t)};
+  const auto &parsedExpr{
+      std::get<parser::Scalar<parser::Expr>>(selectCaseStmt.statement.t).thing};
+  std::optional<TypeCategory> selectCaseStmtType;
+  bool validCaseStmtType{false};
+  if (const auto *expr{GetExpr(parsedExpr)}) {
+    if (auto type{expr->GetType()}) {
+      selectCaseStmtType = type->category();
+    }
+  }
+  if (!selectCaseStmtType ||
+      ((selectCaseStmtType.value() != TypeCategory::Integer) &&
+          (selectCaseStmtType.value() != TypeCategory::Character) &&
+          (selectCaseStmtType.value() != TypeCategory::Logical))) {  // C1145
+    context_.Say(parsedExpr.source,
+        "SELECT CASE expression must be of type CHARACTER, INTERGER, OR LOGICAL"_err_en_US);
+  } else {
+    validCaseStmtType = true;
+  }
+
+  const auto &caseList{
+      std::get<std::list<parser::CaseConstruct::Case>>(selectCaseConstruct.t)};
+  bool defaultCaseFound{false};
+  using var_t = std::variant<std::int64_t, bool, std::string>;
```

Comment:
Don't use `_t` for type names. Can you give this a better name? What does it represent?